### PR TITLE
initial modularisation of 'cpdBootstrap'

### PR DESCRIPTION
**Pull request by j-emberton**
initial modularisation of 'cpdBootstrap'

Currently breaks regression tests - unknown why

Added pytest-xdist to parallelise running of tests (currently set to 1 in pyproject.toml)

### CODE REVIEW

**Comment by j-emberton** [comment]
I've had an initial look at modularising and testing cpdBootstrap.

I think the use of varargin in some functions is breaking the regression tests somehow. The problem is that 'Stats2' can no longer be interrogated properly. However all unit tests pass.

Would appreciate a 2nd pair of eyes on it.

File: oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/get_nt.m [hunk 1]
```diff
@@ -0,0 +1,3 @@
+function [nt] = get_nt(t)
+    nt = length(t);
+end
```
Comment (dorchard) [inline oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/get_nt.m]: Maybe this one is overkill?

File: oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/setup_Stats.m [hunk 2]
```diff
@@ -0,0 +1,17 @@
+function [Stats] = setup_Stats(nBoot, nSeasons, nStrataX, varargin)
+
+    StatsMT = generate_statsMT();
+
+    Stats=StatsMT;
```
Comment (dorchard) [inline oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/setup_Stats.m:5]: This looks like a good transcription - I do wonder what the point of this line is though!

File: tests/conftest.py [hunk 3]
```diff
@@ -272,6 +283,32 @@ def to_matlab_type(data):
...
+
+    # Check if lengths are the same
+    if len(result_list) != len(expected_list):
+        return False
+    
+    # Compare each element
+    for r, e in zip(result_list, expected_list):
+        # Extract scalar value if wrapped in an array-like structure
+        r_value = r[0] if isinstance(r, (list, np.ndarray, matlab.double)) and len(r) > 0 else r
+        e_value = e[0] if isinstance(e, (list, np.ndarray, matlab.double)) and len(e) > 0 else e
```
Comment (dorchard) [inline tests/conftest.py]: do we need to deal with the case where `r` and `e` have size > 1 ?

File: pyproject.toml [hunk 4]
```diff
@@ -1,12 +1,4 @@
...
-addopts = """
-  -v 
-  -p no:warnings
-  --doctest-modules
-  --ignore=tests/test_context.py
-  --show-progress
-  --cov=pyrealm  
-  --cov=ONEFlux/ tests/
-  """
+addopts = "-v -p no:warnings --doctest-modules --ignore=tests/test_context.py --show-progress -n 1 --dist=loadfile"
```
Comment (dorchard) [inline pyproject.toml]: Hm `-n1 --dist=loadfile` doesn't seem to work for me with Python 3.11 version of pytest. I guess my pytest is just a bit old but I thought I was using the latest I could with python 3.11 (which is the latest Python to support the required MATLAB runtime). Is there at least a way of gating this?

Comment (dorchard) [reply pyproject.toml]: yes doesn't work with pytest-8.3.3

**Comment by casterkay** [comment]
> I've had an initial look at modularising and testing cpdBootstrap.
> 
> I think the use of varargin in some functions is breaking the regression tests somehow. The problem is that 'Stats2' can no longer be interrogated properly. However all unit tests pass.
> 
> Would appreciate a 2nd pair of eyes on it.

It turns out that the output `Stats2` is a double string `'"..."'`, so after json decoding it's still a string. 

It's because in `setup_Stats.m` you added 

```
if size(varargin)>0;
    Stats = jsonencode(Stats);
end
```

And in the `cpdBootstrap....m` file there is also another `jsonencode` if `size(varargin) > 0` .

**Comment by dorchard** [comment]
Hm my tests are failing with a lot of `Undefined function 'XXX' for input arguments of type 'int64'` (or `double').

File: oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/setup_Stats.m [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+	    for iBoot=1:nBoot;
+		    for iSeason=1:nSeasons;
+			    for iStrata=1:nStrataX;
+				    Stats(iSeason,iStrata,iBoot)=StatsMT;
+			    end; 
+		    end; 
+	    end;
+	if size(varargin)>0;
+		Stats = jsonencode(Stats);
+	end
```
Comment (casterkay) [inline oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/setup_Stats.m]: If `Stats` is encoded as string at setup, how do you bootstrap it in https://github.com/Cambridge-ICCS/ONEFlux/blob/86931f83ff4da5ed5bb231979eb53d35b5e9f0ba/oneflux_steps/ustar_cp_refactor_wip/cpdBootstrapUStarTh4Season20100901.m#L119?

**Review by casterkay — CHANGES_REQUESTED** [review]

File: oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/setup_Stats.m (hunk 5 above)
Comment (j-emberton) [reply oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/setup_Stats.m]: It only needs to be encoded when we directly test the setupStats function. When that function is called from cpdBootstrap, stats2 shouldn't be encoded as json.

Comment (casterkay) [reply oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/setup_Stats.m]: But there was an additional argument `0` for `setup_Stats` in Line 88-89. I've just removed them.

**Comment by j-emberton** [comment]
Still tests failing after your commit @tztsai , it's just different tests now. Did you run the test suite before you committed your change?

image

**Comment by casterkay** [comment]
The tests were passed on my machine
image

**Comment by j-emberton** [comment]
I run 58 tests, whereas you have only run 52. Have you synced all tests locally?

**Comment by casterkay** [comment]
It's up to date with origin. But I only ran the unit tests. Now there are 58 tests.
image

**Comment by j-emberton** [comment]
> It's up to date with origin. But I only ran the unit tests. Now there are 58 tests. image

Do they all pass?

**Comment by tztsai** [comment]
> > It's up to date with origin. But I only ran the unit tests. Now there are 58 tests. image
> 
> Do they all pass?

It seems so

**Comment by casterkay** [comment]
Some of them failed because of this error. It's due to the `getfield` function and seems unrelated to `Stats`. Others are due to the extra input arguments to `setup_Stats` in the test, so I restored the `varargin` in `setup_Stats`. Current only the dot indexing errors are left.
image

**Comment by j-emberton** [comment]
> Some of them failed because of this error. It's due to the `getfield` function and seems unrelated to `Stats`. Others are due to the extra input arguments to `setup_Stats` in the test, so I restored the `varargin` in `setup_Stats`. Current only the dot indexing errors are left. image

Have you looked at the cpdAssign function? There's a section in it where it parses the stats2 data to extract certain fields. I think that's where it's failing.

File: oneflux_steps/ustar_cp_refactor_wip/launch.m [hunk 6]
```diff
@@ -315,7 +315,7 @@
       
         [Cp2,Stats2,Cp3,Stats3] = ... 
             cpdBootstrapUStarTh4Season20100901 ...
-                (t,NEE,uStar,T,fNight,fPlot,cSiteYr,nBoot); 
+                (t,NEE,uStar,T,fNight,fPlot,cSiteYr,nBoot, 0);
```
Comment (casterkay) [inline oneflux_steps/ustar_cp_refactor_wip/launch.m]: I think there should not be an additional argument here as `Stats` will be a string and this breaks the regression tests.

**Review by casterkay — DISMISSED** [review]

File: oneflux_steps/ustar_cp_refactor_wip/launch.m (hunk 6 above)
Comment (dorchard) [reply oneflux_steps/ustar_cp_refactor_wip/launch.m]: Agreed in dev meeting.

Comment (casterkay) [reply oneflux_steps/ustar_cp_refactor_wip/launch.m]: Fixed in the new commit and the regression tests can pass now

**Comment by j-emberton** [comment]
@dorchard , did you say you had written some comments on this PR. I don't see any here.

**Comment by dorchard** [comment]
> @dorchard , did you say you had written some comments on this PR. I don't see any here.

Sorry forgot to submit the review. Done now.

File: pyproject.toml (hunk 4 above)
Comment (j-emberton) [reply pyproject.toml]: Works for me fine with the same settings. If you have time we could look at it together and try and get to the bottom of this.

File: tests/conftest.py (hunk 3 above)
Comment (j-emberton) [reply tests/conftest.py]: Possibly. Not for this specific part of the project, but there might be as yet undiscovered functions for which it would be useful to have that feature.

How about for the time being I put in a catch to identify any instances where size>1 is True, and exit? If we need that functionality we can add at that point.

Can also write code for n length now if you think it's necessary. Might just need some recursion or something.

File: oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/setup_Stats.m (hunk 2 above)
Comment (j-emberton) [reply oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/setup_Stats.m:5]: Hangover from the previous implementation. I'll fix!

File: oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/get_nt.m (hunk 1 above)
Comment (j-emberton) [reply oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/get_nt.m]: I did um and ah over this but agree it's not strictly necessary! I was hooked on having a nice tidy list of function calls in the main function. I'll remove along with its tests

File: oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/setup_Stats.m (hunk 2 above)
Comment (j-emberton) [reply oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/setup_Stats.m:5]: It turns out that removing the `stats = StatsMT` line changes the behaviour for edge cases where one or more of nBoot, nSeasons or nStrataX are zero.

This doesn't affect the normal running of the code so doesn't trip the regression tests. But it's something we could maybe ask Gilberto what the design intent is?
I'll remove it for now and add comments to highlight this.

File: oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/get_nt.m (hunk 1 above)
Comment (j-emberton) [reply oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/get_nt.m]: fixed

File: oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/setup_Stats.m (hunk 2 above)
Comment (j-emberton) [reply oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/setup_Stats.m:5]: fixed

File: tests/conftest.py (hunk 3 above)
Comment (j-emberton) [reply tests/conftest.py]: fixed with recursion

**Comment by j-emberton** [comment]
all tests pass

File: oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/setup_Stats.m (hunk 2 above)
Comment (dorchard) [reply oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/setup_Stats.m:5]: Okay - perhaps it's better to preserve the edge cases then even if we don't have examples of this now in the test data.

File: tests/conftest.py (hunk 3 above)
Comment (dorchard) [reply tests/conftest.py]: Looks good, marking as resolved.

File: oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/setup_Stats.m (hunk 2 above)
Comment (j-emberton) [reply oneflux_steps/ustar_cp_refactor_wip/cpdBootstrap_modules/setup_Stats.m:5]: OK, I just mentioned this to Gilberto and he also said best to preserve original behaviour, and then we can review in the future. I'll update the code again.

**Review by dorchard — APPROVED** [review]

**Comment by dorchard** [comment]
I think we can merge this now.

**Review by casterkay — APPROVED** [review]
LGTM
